### DISCUSSION TITLE
How to implement an efficient IP access frequency limiting feature?

### DISCUSSION THREAD

**Discussion by EmptyDreams** [vercel/community#3665] [Help] [answered]
### Summary

I would like to limit the frequency of access per IP to prevent malicious attacks (in addition to the fact that I may need to store some data that the server needs to use frequently), and I'd like to know what options are available to do this.

As far as I know it can be done with Vercel KV, but the Vercel KV Hobby plan's daily request limit of 3000 is a bit too low for me. It is also possible to do this with a database such as mongodb, but the performance doesn't seem to be particularly good.

I realise that it may not be particularly good to seek a free solution, but at the moment I have no way of paying for it, so I hope that's understandable.

### Example

none

### Steps to Reproduce

none

**Comment by amyegan** [comment, marked as answer]
Hey @EmptyDreams. Here are a few resources outlining free or low-cost rate limiting solutions:
- [Rate-Limiting Your Serverless Endpoints Without A Database](https://tomichen.com/blog/posts/20210602-rate-limiting-serverless-without-database/)
- [Securing your AI applications with Rate Limiting](https://vercel.com/guides/securing-ai-app-rate-limiting)
- [Rate Limiting Your Next.js App with Vercel Edge](https://upstash.com/blog/edge-rate-limiting)
- [Rate-limit Next.js API routes](https://kittygiraudel.com/2022/05/16/rate-limit-nextjs-api-routes/)

**Reply by EmptyDreams** [reply]
Thank you very much for your help!

Another question I would like to know, is it possible to get the client's IPv4 address by some means on vercel? `request.scoket.remoteAddress` seems like it might return IPv6, but one of the APIs I use only accepts IPv4.
